### PR TITLE
Add Tooltips

### PR DESCRIPTION
**Pull request by psykzz**
## About The Pull Request

This adds Tooltips from /TG/

## Why It's Good For The Game

Better readability for on screen elements

## Changelog
:cl:
add: Tooltips now appear on certain things to help with readability, first pass adds tooltips to on screen actions.
/:cl:

### CODE REVIEW

File: code/modules/tooltip/tooltip.dm [hunk 1]
```diff
@@ -0,0 +1,128 @@
...
+		stuff.send(owner)
+		
+		owner << browse(file2text('code/modules/tooltip/tooltip.html'), "window=[control]")
+
+	..()
+
+
+/datum/tooltip/proc/show(atom/movable/thing, params = null, title = null, content = null, theme = "default", special = "none")
+	if (!thing || !params || (!title && !content) || !owner || !isnum(world.icon_size))
+		return 0
```
Comment (DominikPanic) [inline code/modules/tooltip/tooltip.dm]: stack trace?
```suggestion
		return FALSE
```

File: code/modules/tooltip/tooltip.dm [hunk 2]
```diff
@@ -0,0 +1,128 @@
...
+	if (!init)
+		//Initialize some vars
+		init = 1
```
Comment (DominikPanic) [inline code/modules/tooltip/tooltip.dm]:
```suggestion
		init = TRUE
```

File: code/modules/tooltip/tooltip.dm [hunk 3]
```diff
@@ -0,0 +1,128 @@
...
+- You may have noticed 90% of the work is done via javascript on the client. Gotta save those cycles man.
+- This is entirely untested in any other codebase besides goonstation so I have no idea if it will port nicely. Good luck!
+	- After testing and discussion (Wire, Remie, MrPerson, AnturK) ToolTips are ok and work for /tg/station13
+*/
+
+
+/datum/tooltip
+	var/client/owner
+	var/control = "mainwindow.tooltip"
+	var/showing = 0
```
Comment (DominikPanic) [inline code/modules/tooltip/tooltip.dm]:
```suggestion
	var/showing = FALSE
```

File: code/modules/tooltip/tooltip.dm [hunk 4]
```diff
@@ -0,0 +1,128 @@
...
+	var/queueHide = 0
```
Comment (DominikPanic) [inline code/modules/tooltip/tooltip.dm]:
```suggestion
	var/queueHide = FALSE
```

File: code/modules/tooltip/tooltip.dm [hunk 5]
```diff
@@ -0,0 +1,128 @@
...
+	var/init = 0
```
Comment (DominikPanic) [inline code/modules/tooltip/tooltip.dm]:
```suggestion
	var/init = FALSE
```

File: code/modules/tooltip/tooltip.dm [hunk 6]
```diff
@@ -0,0 +1,130 @@
...
+	if (!thing || !params || (!title && !content) || !owner || !isnum(world.icon_size))
+		return FALSE
+	if (!init)
+		//Initialize some vars
+		init = TRUE
+		owner << output(list2params(list(world.icon_size, control)), "[control]:tooltip.init")
+
+	showing = TRUE
+
+	if (title && content)
```
Comment (Rohesie) [inline code/modules/tooltip/tooltip.dm]:
```
if(title)
	if(content)
		title = "<h1>[title]</h1>"
		content = "<p>[content]</p>"
	else
		title = "<p>[title]</p>"
else if(content)
	content = "<p>[content]</p>"
```

Comment (psykzz) [reply code/modules/tooltip/tooltip.dm]: I'd rather just force a title as h1 and content as p and force the contributor to code it correctly, then we can just do
```
if(title)
    title = "<h1>[title]</h1>"
if(content)
    content = "<p>[content]</p>"
```

Comment (Rohesie) [reply code/modules/tooltip/tooltip.dm]: Up to you.

**Comment by Hughgent** [comment]
If you would be so kind as to make it an toggle option in the preferences tab, that would be nice. Default to on of course.

**Pull request merged**